enterprise/server/ociregistry/ociregistry.go

```diff
@@ -0,0 +1,305 @@
+package ociregistry
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"io"
+	"net/http"
+
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/buildbuddy-io/buildbuddy/server/environment"
+	"github.com/buildbuddy-io/buildbuddy/server/metrics"
+	"github.com/buildbuddy-io/buildbuddy/server/util/log"
+	"github.com/buildbuddy-io/buildbuddy/server/util/prefix"
+	"github.com/buildbuddy-io/buildbuddy/server/util/status"
+	"github.com/google/go-containerregistry/pkg/name"
+	v1 "github.com/google/go-containerregistry/pkg/v1"
+	"github.com/google/go-containerregistry/pkg/v1/remote"
+	"github.com/google/go-containerregistry/pkg/v1/remote/transport"
+	"github.com/google/go-containerregistry/pkg/v1/types"
+)
+
+const (
+	rangeHeaderBytesPrefix = "bytes="
+
+	registryInstanceName = "registry_artifacts"
+)
+
+var (
+	manifestReqRE = regexp.MustCompile("/v2/(.+?)/manifests/(.+)")
+	blobReqRE     = regexp.MustCompile("/v2/(.+?)/blobs/(.+)")
+	blobUploadRE  = regexp.MustCompile("/v2/(.+?)/blobs/uploads/")
+
+	enableRegistry = flag.Bool("ociregistry.enabled", false, "Whether to enable registry services")
+)
+
+type registry struct {
+	env environment.Env
+}
+
+func Register(env environment.Env) error {
+	if !*enableRegistry {
+		return nil
+	}
+
+	mux := env.GetInternalHTTPMux()
+	if mux == nil {
+		return status.FailedPreconditionErrorf("Registry requires internal HTTP mux")
+	}
+
+	r, err := New(env)
+	if err != nil {
+		return err
+	}
+
+	mux.Handle("/v2/", r)
```

Comment:
nit: does the docker HTTP_PROXY support path prefixes? wondering if we could put this under some path like `/oci/registry/v2/` for improved clarity (e.g. when looking at HTTP logs / metrics)

Reply:
The examples in the [OCI Distribution Spec](https://github.com/opencontainers/distribution-spec/blob/main/spec.md)  only show the `/v2/` path, so I'm going to leave it for now.